workflow/scripts/ihw-fdr-control.R

```diff
@@ -0,0 +1,116 @@
+log <- file(snakemake@log[[1]], open="wt")
+sink(log)
+sink(log, type="message")
+
+library("tidyverse")
+library("ggpubr")
+library("IHW")
+
+number_of_groups <- 7
+
+gene_data <- na.omit(read_tsv((snakemake@input[[1]])))
+level_name <- snakemake@wildcards[["level"]]
+
+# calculate covariate mean_obs for gene.aggregated data
+if(level_name == "genes-aggregated") {
+  gene_data <- gene_data %>%
+    mutate(mean_obs = if_else(num_aggregated_transcripts > 0, (sum_mean_obs_counts / num_aggregated_transcripts), 0))
+}
+
+# determine the appropriate number of groups for grouping in ihw calculation for genes_aggregated
+tested_number_of_groups <- number_of_groups
+ihw_test_grouping <- function(x){
+  tryCatch(
+    expr = {
+      ihw(pval ~ mean_obs, data = gene_data, alpha = 0.1, nbins = x)
+      return(TRUE)
+    },
+    error = function(e) {
+      print(str_c("Number of groups was set too hight, trying grouping by ", tested_number_of_groups - 1, " groups."))
+      return(FALSE)
+    })
+}
+
+while(!ihw_test_grouping(tested_number_of_groups) && tested_number_of_groups > 0){
+  tested_number_of_groups <- tested_number_of_groups -1
+  ihw_test_grouping(tested_number_of_groups)
+}
+
+if (tested_number_of_groups < number_of_groups) {
+  print(str_c("The chosen number of groups for ", level_name, " dataset was too large, instead IHW was calculated on ", tested_number_of_groups, " groups."))
+  number_of_groups <- tested_number_of_groups
+}
+
+#select the necessary data
+if(level_name == "genes-aggregated") {
+  gene_data <- gene_data %>%
+    select(-c(qval, sum_mean_obs_counts, num_aggregated_transcripts)) %>%
+    mutate(grouping = groups_by_filter(mean_obs, number_of_groups))
+} else {
+  gene_data <- gene_data %>%
+    select(ens_gene, ext_gene, pval, mean_obs) %>%
+    mutate(grouping = groups_by_filter(mean_obs, number_of_groups))
```

Comment:
I think here I would add another column with the grouping as a factor, for the plotting below, i.e.:
```
    mutate(
      grouping = groups_by_filter(mean_obs, number_of_groups),
      grouping_factor = factor(grouping, ordered = TRUE)
    )
```